--- rdma-core-21/libibverbs/CMakeLists.txt.orig	2018-11-15 01:24:06.000000000 +0100
+++ rdma-core-21/libibverbs/CMakeLists.txt	2018-12-13 18:04:39.219223470 +0100
@@ -76,7 +76,7 @@
     if (NOT NL_KIND EQUAL 0)
       set(REQUIRES "libnl-3.0, libnl-route-3.0")
     endif()
-    rdma_pkg_config("ibverbs" "${REQUIRES}" "${PROVIDER_LIBS} -libverbs ${CMAKE_THREAD_LIBS_INIT}")
+    rdma_pkg_config("ibverbs" "${REQUIRES}" "-libverbs ${CMAKE_THREAD_LIBS_INIT}")
 
     file(WRITE ${BUILD_INCLUDE}/infiniband/all_providers.h "#define FOR_EACH_PROVIDER() ${FOR_EACH_PROVIDER}")
   else()
